Standards/scs-0216-v1-high-security-setup.md

```diff
@@ -0,0 +1,166 @@
+---
+title: Kubernetes high security setup
+type: Standard
+status: Draft
+track: KaaS
+---
+
+## Introduction
+
+A Kubernetes instance is provided as a cluster, which consists of a set of machines,
+so-called nodes. These clusters are configured depending on the needs of its users, which
+is possible due to the flexibility of Kubernetes.
+Due to the regular changes, there are always new security features to deploy and use in Kubernetes.
+Nevertheless, a provider (or even a customer) needs to take action in order to achieve a
+hardened, secure cluster due to the myriad of configurations possible. This is especially
+the case since Kubernetes ships with insecure features and configurations out of the box,
+which will need to be mitigated by an administrator with the proper knowledge.
+Hardened, secure Kubernetes clusters are desirable regardless of the possible threat model,
+since higher security doesn't necessarily mean higher complexity in this case.
```

Comment:
As far as recommendations in this document go, I strongly agree with that.

However, again, as far as recommendations in this document go, I'd probably call it rather a security baseline standard, while referencing primarily to CIS benchmarking.

Going further, a "High Security Kubernetes Cluster" would probably include points like...
- network protection of the API server using IP allow lists or using bastion hosts
- support for CRI runtimes supporting VM-based runtime isolation
- disabling generation of admin client certificates handed to users
- ...

Reply:
Since I only wrote this standard with the points given by the issue and I'm not that knowledgeable in this specific field (and also not with the points you bring up), I would probably rename the standard to "K8s Security Baseline" and put the points you mentioned in a new standard.

Also just to ask: Do we trust CIS as a standard enough to put it as a reference to such a big topic? They probably do good work, but we also would need to review their standards (versions) constantly to ensure consistency with our requirements.

Reply:
> Since I only wrote this standard with the points given by the issue and I'm not that knowledgeable in this specific field (and also not with the points you bring up), I would probably rename the standard to "K8s Security Baseline" and put the points you mentioned in a new standard.

:+1: 

[Your comment from above:](https://github.com/SovereignCloudStack/standards/pull/376#discussion_r1403370342)
> 
> Do you have a specific document of them in mind (TBF I'm not willing to give them my mail, even if its just my work mail). Then I can/will rewrite the text to mention these standards.

I found the access to the PDF to be gated through a registration to be odd, as well (https://www.cisecurity.org/benchmark/kubernetes). Also did not complete the form due to concerns that I might violate their TOS by getting their PDF to use it kind-of commercially without license.
Definitely, it appears to be an de-facto industry standard; [Azure](https://learn.microsoft.com/en-us/azure/aks/cis-kubernetes), [Google](https://cloud.google.com/kubernetes-engine/docs/concepts/cis-benchmarks#status), [VMware](https://docs.vmware.com/en/VMware-Tanzu-Mission-Control/services/tanzumc-concepts/GUID-1BB98612-A9BF-494C-8446-1DB2E80BF5F9.html) and probably others use it to benchmark their offerings. Of course, they also may have a business license.

While concerns around this topic are IMHO valid, I concluded that they are negligible, as AFAIK kube-bench is Apache-licensed and we AFAIK can read it with any (non-)commercial intent. That is, of course, given kube-bench maintainers had the rights to publish it like that to begin with :D

> Also just to ask: Do we trust CIS as a standard enough to put it as a reference to such a big topic? They probably do good work, but we also would need to review their standards (versions) constantly to ensure consistency with our requirements.

That is a good question indeed. Especially as the CIS benchmarks do not seem to be really free/open that it would be a great match with SCS.

But I do not think that there are many viable alternatives. E. g. if SCS-employed/contracted staff was tasked with writing up a similar security benchmark, they would...

- ...need a fair amount of time
- ...miss quite a few points OR mostly copy/paste from CIS benchmarks

None of these outcomes would prevent future efforts to keep them up-to-date OR make the benchmark both high-quality and completely free/open.

Reply:
Makes a lot sense to me, thank you for the detailed explanation.
I think a good way to (at least) get on top of these standards would be to reference a specific version in one of our standards. If we agree with this version (if they version their standards, which they hopefully do), we would have a baseline, that we could later expand on if we find the time to read a newer version of it and agree on its contents.
I'm still not happy with the gate process and we should probably look into the license problem before we continue down this path, but at least that sounds good to me.

Reply:
Did a bit more reading up. The CIS benchmarks themselves are available under a [CC BY-NC-SA 4.0 Deed ](https://creativecommons.org/licenses/by-nc-sa/4.0/deed.en) license, meaning that we could use it as long as it is for non-commercial use (which I think is the case).

The Gating process seems to be in place in order to enforce their ToS, which mainly say that they're not responsible for anything done with the help of/under their tests (so they're not liable).

While I agree with this, I'm still unhappy with the gating process.
For this reason and because I won't wait for a decision on this (and maybe someone else has more insight into this to write something useful), I would drop the parts mentioned in https://github.com/SovereignCloudStack/standards/pull/376#discussion_r1397322690 and publish the document without them. We can then create a Version 2 of this standard, which would include the CIS benchmark mentions.

I think thats the best solution for now, otherwise we would sit endlessly on this issue.

Reply:
> I would keep the parts mentioned in #376 (comment), since they still provide a good foundation, and publish the document in its current state. We can then create a Version 2 of this standard, which would include the CIS benchmark mentions and maybe drop the parts mentioned.

Alright!

> For this reason and because I won't wait for a decision on this

Going forward (v2 and beyond), it probably is a good idea to identify someone with a fitting role to make such decision. I guess someone from the core SCS project team.

Reply:
Thank you both for your work! 

The CIS benchmark covers a lot of security configuration options, for every component of Kubernetes. This would definitely be too much for a SCS standard document. Referring to it is the way to go. 

This also keeps the standard up-to-date when the K8s version changes and security needs to be adapted. New releases of the CIS benchmark then covers it automatically.